### PR TITLE
dice_loss cleanup, and adding MaskedDiceLoss

### PR DESCRIPTION
**Pull request by myron**
1. dice_loss cleanup - no changes (just some code cleanup)

2. MaskedDiceLoss - similar to DiceLoss but accepts the mask over which to compute. It's currently a sublclass of DiceLoss, but could be made a part of DiceLoss.  No unit tests added.

### CODE REVIEW

File: monai/losses/dice.py [hunk 1]
```diff
@@ -84,41 +94,75 @@ def forward(self, input: torch.Tensor, target: torch.Tensor, smooth: float = 1e-
...
+            f = torch.sum(f)  # sum over the batch and channel dims
+        elif self.reduction == "none":
+            pass  # returns [N, n_classes] losses
+        else:
+            raise ValueError(f"reduction={self.reduction} is invalid.")
+
+        return f
+
+
+class MaskedDiceLoss(DiceLoss):
```
Comment (wyli) [inline monai/losses/dice.py:139]: This introduces new features, could you add a unit test?

File: monai/losses/dice.py [hunk 2]
```diff
@@ -84,41 +94,75 @@ def forward(self, input: torch.Tensor, target: torch.Tensor, smooth: float = 1e-
...
+    """
+    Same as DiceLoss, but accepts a binary mask ([0,1]) indicating a region over which to compute the dice.
+    """
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor, mask: torch.Tensor = None, smooth: float = 1e-5):
+        """
+        Args:
+            input (tensor): the shape should be BNH[WD].
+            target (tensor): the shape should be BNH[WD].
+            mask (tensor): (optional) the shape should B1H[WD]
```
Comment (wyli) [inline monai/losses/dice.py:149]: I think the `mask` supports the broadcasting rules by construction, such as  `BNH[WD]`, `BN1[, 1, 1]`, `BNH[W, 1]`... the docstring needs to be updated

File: tests/test_dice_loss.py [hunk 3]
```diff
@@ -114,13 +114,13 @@
 class TestDiceLoss(unittest.TestCase):
     @parameterized.expand(TEST_CASES)
     def test_shape(self, input_param, input_data, expected_val):
-        result = DiceLoss(**input_param).forward(**input_data)
+        result = DiceLoss(**input_param)(**input_data)
```
Comment (wyli) [inline tests/test_dice_loss.py:117]: this actually prevents potential bugs, thanks!

File: monai/losses/dice.py [hunk 4]
```diff
@@ -84,41 +94,75 @@ def forward(self, input: torch.Tensor, target: torch.Tensor, smooth: float = 1e-
...
+            pass  # returns [N, n_classes] losses
+        else:
+            raise ValueError(f"reduction={self.reduction} is invalid.")
+
+        return f
+
+
+class MaskedDiceLoss(DiceLoss):
+    """
+    Same as DiceLoss, but accepts a binary mask ([0,1]) indicating a region over which to compute the dice.
```
Comment (wyli) [inline monai/losses/dice.py:141]: looks like the implementation supports non-binary masks as well. this docstring should be updated

File: monai/losses/dice.py (hunk 1 above)
Comment (myron) [reply monai/losses/dice.py:139]: I added unit tests

File: monai/losses/dice.py (hunk 2 above)
Comment (myron) [reply monai/losses/dice.py]: we'll support only B1H[WD] and BNH[WD]. added text

File: tests/test_dice_loss.py (hunk 3 above)
Comment (myron) [reply tests/test_dice_loss.py]: I changed this back to have .forward

File: monai/losses/dice.py (hunk 4 above)
Comment (myron) [reply monai/losses/dice.py:141]: it's designed for binary masks only. but won't be checking this, since it's an overhead

**Comment by myron** [comment]
done
